hnix-store-core/src/System/Nix/Internal/Path.hs

```diff
@@ -0,0 +1,150 @@
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE TypeApplications           #-}
+
+{-|
+Description : Internal path utilities
+Maintainer  : srk <srk@48.io>
+|-}
+module System.Nix.Internal.Path where
+
+import           Control.Monad
+import qualified Data.ByteString.Lazy      as BSL
+import qualified Data.ByteString.Char8     as BSC
+import           Data.Text                 (Text)
+import qualified Data.Text                 as T
+import           System.FilePath.Posix     (takeBaseName, takeDirectory)
+import           System.Nix.Path           (Path(..), PathName(..), pathName, PathHashAlgo)
+import           System.Nix.Internal.Hash  (Digest(..),  HashAlgorithm'( SHA256 ))
+import           System.Nix.Hash           (HasDigest, printAsBase16, printAsBase32, printHashAlgo)
+import qualified System.Nix.Hash
+
+-- | Parse store location
+parseStore :: BSL.ByteString -> T.Text
+parseStore = T.pack . takeDirectory . BSC.unpack . BSL.toStrict
+
+-- | Parse path from string
+--
+-- in form <storeDir>/<hash>-<pathName>
+-- into (Just (Path <hash> (PathName <pathName>)))
+-- or Nothing on error
+--
+-- XXX: should check for @PathHashAlgo length
+parsePath :: BSL.ByteString -> Maybe Path
+parsePath p = case name of
+    Nothing -> Nothing
+    Just n -> Just $ Path digest n
+  where
+  base = T.pack . takeBaseName . BSC.unpack . BSL.toStrict $ p
+  parts = T.breakOn "-" base
+  digest = Digest . BSC.pack . T.unpack . fst $ parts
+  name = pathName . T.drop 1 . snd $ parts
+
+
+-- experimental
+-- Directory of the store
+type StoreDir = Text
+type Stored a = (StoreDir, a)
+
+-- wrap StoreDir and Path into tuple
+makeStored :: StoreDir -> Path -> Stored Path
+makeStored sl p = (sl, p)
+
+type PathType = Text
```

Comment:
Could we create a new datatype for this? I think it would be:
```
-- | Store object type is used when creating store paths, to differentiate
-- between paths for sources and derivation outputs (see thesis page 94)
data StoreObjectType =
    Source
  | Output Text -- ^ A build output (e.g. `Output "out"`)
```
or
```
-- | A single derivation build output (e.g. "out", "docs" etc) 
newtype BuildOutputName = BuildOutputName { getBuildOutputName :: Text }

-- | Store object type is used when creating store paths, to differentiate
-- between paths for sources and derivation outputs (see thesis page 94)
data StoreObjectType =
    Source
  | Output BuildOutputName -- ^ A build output (e.g. `Output (BuildOutputName "out")`)
```